### DISCUSSION TITLE
Error while building project.

### DISCUSSION THREAD

**Discussion by jkugsiya** [oppia/oppia-android#4984] [Q&A (Installation)]
### Operating System

Windows 11

### Android Studio Version

Android Studio Flamingo | 2022.2.1 Patch 1

### RAM size

16 GB

### Free disk space

300 GB

### Which step are you stuck on?

I am getting this error:
![image](https://github.com/oppia/oppia-android/assets/42642325/affbdcb9-fa47-4906-845e-ea80054efcfb)

I've tried the steps found in a GitHub discussion to change encoding to UTF-8 but not solved yet.
![image](https://github.com/oppia/oppia-android/assets/42642325/017b72dc-2d10-448a-95d5-1a3bd1599fce)

### Error Log

```bash
D:\Projects\Android\oppia-android\model\build\generated\source\proto\main\java\org\oppia\android\app\model\OppiaLanguage.java:25: error: unmappable character for encoding Cp1252
   * Corresponds to the Arabic (اَلْعَرَب�?يَّة�?‎) macro language. IETF BCP 47 language tag: ar.
                                                   ^
D:\Projects\Android\oppia-android\model\build\generated\source\proto\main\java\org\oppia\android\app\model\OppiaLanguage.java:25: error: unmappable character for encoding Cp1252
   * Corresponds to the Arabic (اَلْعَرَب�?يَّة�?‎) macro language. IETF BCP 47 language tag: ar.
                                                             ^
D:\Projects\Android\oppia-android\model\build\generated\source\proto\main\java\org\oppia\android\app\model\OppiaLanguage.java:41: error: unmappable character for encoding Cp1252
   * Corresponds to the Hindi (हिन�?दी) macro language. IETF BCP 47 language tag: hi.
                                          ^
D:\Projects\Android\oppia-android\model\build\generated\source\proto\main\java\org\oppia\android\app\model\OppiaLanguage.java:92: error: unmappable character for encoding Cp1252
   * Corresponds to the Arabic (اَلْعَرَب�?يَّة�?‎) macro language. IETF BCP 47 language tag: ar.
                                                   ^
D:\Projects\Android\oppia-android\model\build\generated\source\proto\main\java\org\oppia\android\app\model\OppiaLanguage.java:92: error: unmappable character for encoding Cp1252
   * Corresponds to the Arabic (اَلْعَرَب�?يَّة�?‎) macro language. IETF BCP 47 language tag: ar.
                                                             ^
D:\Projects\Android\oppia-android\model\build\generated\source\proto\main\java\org\oppia\android\app\model\OppiaLanguage.java:108: error: unmappable character for encoding Cp1252
   * Corresponds to the Hindi (हिन�?दी) macro language. IETF BCP 47 language tag: hi.
                                          ^
```

### Approaches already used to resolve the issue

Tried the solution that was found in a GitHub discussion.

**Comment by MohitGupta121** [comment]
@jkugsiya please follow this:

In Android Studio, the default file encoding can be determined by the operating system's default encoding. Android Studio follows the default encoding set by the underlying platform to handle file encoding for various operations, such as reading and writing files.

On different operating systems, the default file encoding may vary:

Windows: The default file encoding on Windows is typically Cp1252 (also known as Windows-1252).

macOS: The default file encoding on macOS is usually UTF-8.

Linux: On Linux distributions, the default file encoding can vary depending on the distribution and system configuration. However, UTF-8 is widely used and often set as the default encoding.

You can change your file encoding to UTF-8 by visiting `Android Studio Settings > Editor > File Encodings`

<img width="1080" alt="Screenshot 2023-05-15 at 13 31 09" src="https://github.com/oppia/oppia-android/assets/59600948/878eee67-07b6-4fb9-8101-c1c98e931917">


Hope this will fix your issue, if it then please mark this as answer to help others in community.
Thanks!

**Reply by jkugsiya** [reply]
I've already tried this fix and mentioned it as well. But this is not working

**Reply by adhiamboperes** [reply]
@jkugsiya, curious whether you tried an invalidate and rebuild after switching to UTF-8?

**Reply by jkugsiya** [reply]
Yes, tried that as well

**Comment by BenHenning** [comment]
I think this is a problem with your Git client on Windows. It seems that it's automatically converting UTF-8 files coming from GitHub into CP1252 which is tripping up the Java compiler in this case.

For more details: some of the comments in the OppiaLanguage proto file includes Unicode characters:

https://github.com/oppia/oppia-android/blob/da80fab4a78997cb8ab190292f0676117f2722fc/model/src/main/proto/languages.proto#L8C1-L33

These files should be encoded in UTF-8 by default. When the proto compiler generates Java code, it directly copies proto field arguments from the proto file to the generated Java source file. I suspect this is a byte-level copy with no re-encoding going on, so if your Git client is automatically re-encoding pulled UTF-8 files with Unicode characters to CP1252 then we end up with a Java file that's probably in both CP1252 (for the copied comments) and either UTF-8 or ANSI (depending on the proto compiler), and the Java compiler doesn't seem to handle that well.

You may need to configure your local environment to use UTF-8 by default. However, per https://stackoverflow.com/a/48908993/3689782 it seems that the easiest way to do this would actually be by using .gitattributes (which we can check in directly to Oppia Android). A short-term solution may be to forcibly convert all of your checked out repository files to UTF-8 by doing something like: https://stackoverflow.com/a/52695560/3689782.

**Reply by jkugsiya** [reply]
After recloning and changing to utf-8 format, now I am getting this error:

![image](https://github.com/oppia/oppia-android/assets/42642325/18f8df3c-d9d1-4c8a-a5a7-d0caeb17dd95)

**Reply by MohitGupta121** [reply]
This issue is related to our use of jetifier currently. I believe it will be resolved with the upcoming build platform changes, but to explain a bit:

In gradle.properties, `android.enableJetifier=true` automatically converts third-party libraries to use AndroidX. 
The maximum supported major version for jetifier class file format in Android is 58, which corresponds to Java 14.
The `model.jar` was compiled with Java 15/major version 59, hence the incompatibility.

A fix would be to compile the JAR file with a lower version of Java. 

@jkugsiya can you please check using https://developer.android.com/studio/intro/studio-config#jdk.

looks like you're using JDK 15--this isn't supported with our version of Gradle. You'll need to install JDK 8 LTS for it to work correctly (9 is fine too).

**Reply by yimello908** [reply]
Thanks, this help me too

**Reply by MohitGupta121** [reply]
@yimello908 glad to gear that, please mark the answer as "Mark as answer" so that it's help others facing same issue. Thanks!

**Comment by adhiamboperes** [comment]
@Tejas-67

Please see this discussion for https://github.com/oppia/oppia-android/issues/4760#issuecomment-1771012082:

Which Android Studio and Java version do you have? It's recommended to build with Java 11, which is embedded in AS. Please see [setup instructions](https://github.com/oppia/oppia-android/wiki/Installing-Oppia-Android#install-oppia-android) if you need help configuring this.

**Comment by Tejas-67** [comment]
@adhiamboperes  I am using AS Flamingo, Android SDK 31, 28 and Gradle JDK 11. Am i doing something wrong?

PS: the previous encoding issue is fixed but there is a new issue ->

 :model:compileJava failed , :app:processDebugMainfest fail. 
Cause: Cause: class org.gradle.internal.compiler.java.ClassNameCollector (in unnamed module @0xb34acdc) cannot access class com.sun.tools.javac.code.Symbol$TypeSymbol (in module jdk.compiler) because module jdk.compiler does not export com.sun.tools.javac.code to unnamed module @0xb34acdc

**Reply by MohitGupta121** [reply]
@Tejas-67 we apologize, could you please downgrade your Android Studio **Bumblebee** as per [wiki instructions](https://github.com/oppia/oppia-android/wiki/Installing-Oppia-Android#prepare-developer-environment)? Thanks!

**Comment by Tejas-67** [comment]
@jkugsiya how did you fix the " Execution failed for task ':utility:kaptDebugKotlin'. " error ?
Tried setting gradle jdk to jdk 8 and also tried to run in bumblebee as per @MohitGupta121 but still issue not resolved.

**Reply by adhiamboperes** [reply]
It would be really helpful to see a full stack trace so we can pinpoint the actual error.

Also, cleaning your gradle build would help in case the files are cached

**Reply by Tejas-67** [reply]
@adhiamboperes 

Operating System
Windows 11

Android Studio Version
Android Studio BumbleBee

RAM size
8 GB

ERROR
![image](https://github.com/oppia/oppia-android/assets/110051718/1f09b606-5189-4a3b-a3e6-7a9d907aead2)

**Reply by adhiamboperes** [reply]
I've noticed you made changes to gradle files, could you have modified something that caused an incompatibility? for example if you changed the compileSdkVersion, or the gradle wrapper, you will most likely encounter an error. Does (https://stackoverflow.com/a/69250502/11396524) help?

**Reply by Tejas-67** [reply]
@adhiamboperes yes it worked for me. I have changed org.gradle.jvmargs in gradle.properties from -Xmx4096m to 
-Dfile.encoding=UTF-8 -Xmx4096m to solve the unmappable character error. Will this change affect any other aspect of the application?

**Reply by adhiamboperes** [reply]
Thanks for sharing your solution, we'll investigate if this is a config that should be added to the project.

If the build works on your local, that's okay. You will just gitignore it or keep it out of your commits so that it does not affect other developers at the moment.
